### PR TITLE
Extract minimal Rialto-Sub and Millau-Sub clients

### PR DESCRIPTION
**Pull request by svyatonik**
Code in `relays/ethereum` is now specific to eth2sub && sub2eth, where only supported Sub is currently Rialto-Sub chain (we may add support for other chains as well, but it isn't what I want to do now).
Code in `relays/substrate-client` is a minimal `Client<Chain>` - i.e. it is client that is able to talk with selected Substrate-based chains. There are two builtin chains there (activated with cargo features) - millau and rialto.

### CODE REVIEW

File: Cargo.lock
Comment (svyatonik) [inline Cargo.lock:2615]: There's another small change in our fork of jsonrpsee. [This commit](https://github.com/svyatonik/jsonrpsee/commit/bc0d938217fd0c34e8dcaa08d034ab157c87b363) allows specifying using associated types in RPC methods. I.e.:
```rust
jsonrpsee::rpc_api! {
	pub(crate) Substrate<C: Chain> {
		#[rpc(method = "chain_getHeader", positional_params)]
		fn chain_get_header(block_hash: Option<C::Hash>) -> C::Header;
        }
}
```

Without it, I had to specify all used types explicitly, instead of using 'types container' (`Chain`).

**Comment by svyatonik** [comment]
Next PR will be about relay itself (sub2sub specifically). I kindly ask you to leave your opinion about future of `ethereum-poa-relay` relay here. The possibly options I consider are:
1) rename it to generic relay and use it to serve all sub2eth, eth2sub and sub2sub bridges. I don't think putting everything in a single binary is a best option - e.g. we won't have message-lanes in sub2eth and we do not want (at least initially) currency exchange in sub2sub;
2) rename it to generic relay and turn on/off support for chains/applications using cargo features. Imo it is even worse than (1);
3) add separate relay for serving Millau headers/messages relay (Millau-Sub <-> Rialto-Sub). I guess later we may extend this binary to support other chains as well - but this is yet to be decided.

File: relays/ethereum/src/instances.rs
Comment (tomusdrw) [inline relays/ethereum/src/instances.rs:49]: The `pallet_call` seems identical for `TestPoA` and `Kovan` afaict, can't we extract it?

File: relays/ethereum/src/instances.rs
Comment (tomusdrw) [inline relays/ethereum/src/instances.rs:37]: The reason against `enum` was that it's not extensible - i.e. it's not easy for an external party to add it's own rialto-like runtime with more instances of the bridge.
IMHO ideally the runtime would be generic on the `sync_loop` level, so that other chains can use the library and initialize the event loop with a different runtime.
That said, since we have the runtime hardcoded already I believe having this enum doesn't matter that much. I'd leave it up to you - I'm fine with just creating an issue for this.

File: relays/ethereum/src/instances.rs
Comment (tomusdrw) [inline relays/ethereum/src/instances.rs:85]: Same here, the `pallet_call` seems the duplicated.

File: relays/ethereum/src/ethereum_deploy_contract.rs
Comment (tomusdrw) [inline relays/ethereum/src/ethereum_deploy_contract.rs:107]: Why not keep all the types in a single type (I'm thinking `Rialto` which implements `Chain`)? That would simplfy migration to a fully generic code in the future.

File: relays/substrate-client/src/lib.rs
Comment (tomusdrw) [inline relays/substrate-client/src/lib.rs:40]: Isn't it possible to split this into it's own crates? Instead of activating a feature, the crate would simply have two deps: `substrate-client` and the runtime it uses. If the crate is generic it could just import `substrate-client` and then accept any `Chain` object.
Mid-term I think it would make sure we don't hardcode even more runtime-specific things into the relay code.

File: relays/substrate-client/src/rpc.rs
Comment (tomusdrw) [inline relays/substrate-client/src/rpc.rs:21]: I supposed it's only in the case where none of the features is enabled (i.e. we are just building the crate from the workspace). Splitting to crates, as suggested in another comment, should solve that.

Comment (svyatonik) [reply relays/substrate-client/src/rpc.rs:21]: No, it's for some internal enum that is created by jsonrpsee::rpc_api macro

File: relays/substrate-client/src/lib.rs
Comment (svyatonik) [reply relays/substrate-client/src/lib.rs:40]: It is possible - I just wanted to avoid too many crates. So idea is that we have basic crate with basic definitions + for basic chain that we use. Everyone else may implement their own `trait Chain` to get `Client` functionality. Happy to split it if you think it's better.

> Mid-term I think it would make sure we don't hardcode even more runtime-specific things into the relay code.

I don't get it - how split will help? :)

File: relays/ethereum/src/instances.rs
Comment (svyatonik) [reply relays/ethereum/src/instances.rs:37]: I will try to revert this change - I can't remember why exactly I had to change that :)

Comment (svyatonik) [reply relays/ethereum/src/instances.rs:37]: Having completely runtime-agnostic (you have similar concerns here and in other comments as well) relay is possible - it's just not a part of this PR. I've extracted minimal client mostly to avoid duplicating RPC code.

Even though we have had `trait BridgeInstance`, it was returning `rialto_runtime::Call` :) So having enum imo is the same level of extensibility :) But as I said I'll try to revert this change - I'm not sure why I need to do that :)

File: relays/ethereum/src/ethereum_deploy_contract.rs
Comment (svyatonik) [reply relays/ethereum/src/ethereum_deploy_contract.rs:107]: Most of required types are already there. This PR isn't about making existing relay generic - it is about making some RPC-related code generic so that we may reuse the same code in sub2sub. We have separate issue #61 for making this relay more or less generic.

File: relays/ethereum/src/instances.rs
Comment (svyatonik) [reply relays/ethereum/src/instances.rs]: Reverted

File: relays/ethereum/src/instances.rs
Comment (svyatonik) [reply relays/ethereum/src/instances.rs]: Reverted

File: relays/ethereum/src/instances.rs
Comment (svyatonik) [reply relays/ethereum/src/instances.rs]: Reverted.

File: relays/ethereum/src/ethereum_sync_loop.rs
Comment (HCastano) [inline relays/ethereum/src/ethereum_sync_loop.rs:167]:
```suggestion
	/// Bridge instance used in Ethereum to Substrate sync.
```

**Review by HCastano — COMMENTED** [review]
I'm gonna take another look tomorrow and try running the relay with the new changes. I've got mixed feelings about how some of the stuff got moved and renamed, but I need to think about it a bit more before I give any intelligent feedback.

I think that it would be okay to have two relay binaries: `ethereum-relay` and `substrate-relay`. The Ethereum relay would be used for Eth2Sub, Sub2Eth and ideally support different instances. The Substrate relay would be for Sub2Sub and be configurable again through some instances flag.

**Comment by HCastano** [comment]
> There's another small change in our fork of jsonrpsee. This commit allows specifying using associated types in RPC methods. I.e.:
> 
> jsonrpsee::rpc_api! {
> 	pub(crate) Substrate<C: Chain> {
> 		#[rpc(method = "chain_getHeader", positional_params)]
> 		fn chain_get_header(block_hash: Option<C::Hash>) -> C::Header;
>         }
> }
> 
> Without it, I had to specify all used types explicitly, instead of using 'types container' (Chain).

Is this something that should/can be merged upstream?

File: relays/ethereum/src/instances.rs
Comment (svyatonik) [reply relays/ethereum/src/instances.rs:47]: To avoid confusion - there's type with the same `substrate-client`.

**Comment by svyatonik** [comment]
> Is this something that should/can be merged upstream?

No

File: relays/ethereum/src/instances.rs
Comment (HCastano) [reply relays/ethereum/src/instances.rs:47]: I'd prefer to change the type in `substrate-client` to `RialtoRuntime` and keep this instance type as `Rialto`

**Comment by HCastano** [comment]
> > Is this something that should/can be merged upstream?
> 
> No

Why not?

**Comment by svyatonik** [comment]
> Why not?

In general, yes - I want this functionality in `jsonrpsee` or whatever we want to use. In practice: (1) I haven't checked if there any caveats - I've just changed macro to work in our case (2) this is change that's made on top of other change that we don't want to backport (3) jsonrpsee seems to have no owner (aka unmaintained) atm.

File: relays/ethereum/src/instances.rs
Comment (svyatonik) [reply relays/ethereum/src/instances.rs:47]: I'll change this to RialtoPoA for now - if you'll decide to use something else, it could be changed. So now (in near future) we'll have 3 chains:
- Millau-Sub - chain that has all modules for sub2sub sync with Rialto-Sub;
- Rialto-Sub - chain that has all modules for sub2sub sync with Millau-Sub and all modules to sync with Rialto-PoA;
- Rialto-PoA - Etherum PoA chain.

For simplicity, I suggest to use Millau/Rialto both to identify bridges and their 'main' substrate-based chains. Other chains should be referenced with some suffix - e.g. RialtoPoA. Or we may use some names as you've suggested, but it isn't a part of this PR.

**Review by tomusdrw — APPROVED** [review]
lgtm, although would still prefer crates than conditional compilation.

File: relays/substrate-client/src/lib.rs
Comment (svyatonik) [reply relays/substrate-client/src/lib.rs:40]: > To make sure everything actually works properly you should run multiple separate builds with different combination of feature flags

Ok, this makes sense - I'll extract to separate crates then

File: primitives/ethereum-poa/src/lib.rs
Comment (tomusdrw) [inline primitives/ethereum-poa/src/lib.rs:552]: Luckily we don't have the bridge deployed anywhere :) BTW why do we hardcode the bridge name in the runtime API? This feels like it could be something more generic like: "PoAHeaderApi" - just curious, no changes required here.

File: relays/substrate-client/Cargo.toml
Comment (tomusdrw) [reply relays/substrate-client/Cargo.toml]: Until we have an automated check & formatting tool for this, I'm reluctant to address that. We have `cargo fmt` to avoid biases towards formatting in `.rs` files, and I really think we should do that for `.toml` as well, otherwise it's pretty futile.
Added an issue for this: https://github.com/paritytech/parity-bridges-common/issues/378

File: relays/substrate-client/src/lib.rs
Comment (svyatonik) [reply relays/substrate-client/src/lib.rs]: done

File: relays/substrate-client/src/chain.rs
Comment (tomusdrw) [inline relays/substrate-client/src/chain.rs:34]: It seems that some of these bounds are already part of `Parameter` or `Member`. Do we really need them explicitly?

File: relays/substrate-client/src/client.rs
Comment (tomusdrw) [inline relays/substrate-client/src/client.rs:44]: Why can't we derive it and enforce `Chain` to have `Debug` as well?

File: relays/substrate-client/src/client.rs
Comment (tomusdrw) [inline relays/substrate-client/src/client.rs:43]: I don't think this crate is `no_std` compatible, don't see a feature like that.
```suggestion
```

**Review by tomusdrw — APPROVED** [review]

File: relays/substrate-client/src/chain.rs
Comment (svyatonik) [reply relays/substrate-client/src/chain.rs:34]: It's a plain copy of `frame_system::Trait`. Actually we don't need all of these trait, but will probably need some of these later.

File: primitives/ethereum-poa/src/lib.rs
Comment (svyatonik) [reply primitives/ethereum-poa/src/lib.rs:552]: Because: (1) we have two instances of PoA bridge module in the runtime (2) API is implemented for runtime => we can't implement the same API twice => we need to include instance name into API.

File: relays/substrate-client/src/client.rs
Comment (svyatonik) [reply relays/substrate-client/src/client.rs:44]: `jsonrpsee::Client` is not `Debug`

**Review by HCastano — APPROVED** [review]
Looks good! I like the crates more than having things feature flagged. The naming thing I brought up can definitely be changed later.

**Comment by HCastano** [comment]
CI failures related to #379. I don't want to be too hasty with the merge, so I'd rather wait until we fix our CI to get this in.

**Pull request merged**
